File: lection04/app_java/src/main/java/ru/hse/lection04/businesslayer/AbstractCallbackProvider.java
package ru.hse.lection04.businesslayer;

import java.util.HashSet;
import java.util.Set;

public class AbstractCallbackProvider<LISTENER> {
    protected static final int EMPTY = 0;

    protected final Set<LISTENER> mListeners = new HashSet<>();


    public void register(LISTENER listener) {
        final int lastSize = mListeners.size();

        mListeners.add(listener);

        if (lastSize == EMPTY) {
            setActivation(true);
        }
    }

    public void unregister(LISTENER listener) {
        mListeners.remove(listener);

        if (mListeners.size() == EMPTY) {
            setActivation(false);
        }
    }


    protected void setActivation(boolean value) {

    }
}

File: lection04/app_kotlin/src/main/java/ru/hse/lection04/businesslayer/AbstractCallbackProvider.kt
package ru.hse.lection04.businesslayer


/**
 * Класс, которые содержит логику для работу с подписчиками
 * @param <LISTENER> любой подписчик
 */
open class AbstractCallbackProvider<LISTENER> {
    companion object {
        protected const val EMPTY = 0
    }


    protected val listeners = mutableSetOf<LISTENER>()


    /***
     * Зарегестрировать подписчика
     * @param listener подписчик
     */
    fun register(listener: LISTENER) {
        val lastSize = listeners.size
        listeners.add(listener)
        if (lastSize == EMPTY) {
            setActivation(true)
        }
    }

    /**
     * Удалить подписчика
     * @param listener подписчик
     */
    fun unregister(listener: LISTENER) {
        listeners.remove(listener)
        if (listeners.size == EMPTY) {
            setActivation(false)
        }
    }

    /**
     * В зависимости от изменения набора подписчиков вызовется этот метод
     * @param value true - означает что появились слушатели и можно начать процесс отслеживания даных и уведомления. false - когда список подписчиков пуст
     */
    protected open fun setActivation(value: Boolean) {}
}